[Automation] java run fails for com.lihaoyi:upickle_3:4.4.0

## java run on ubuntu-latest / JDK 25 / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: 25
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.4.0" ./gradlew clean javaTest -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25352097855/job/74333935744
Last 300 lines of the log:
```console
                    at app//upickle.implicits.ReadersVersionSpecific$CaseClassReader3V2$ObjectContext.visitEnd(Readers.scala:69)
                    at app//upickle.implicits.ReadersVersionSpecific$CaseClassReader3V2$ObjectContext.visitEnd$(Readers.scala:27)
                    at app//upickle.implicits.ReadersVersionSpecific$$anon$1.visitEnd(Readers.scala:77)
                    at app//upickle.core.TraceVisitor$$anon$1.visitEnd(TraceVisitor.scala:122)
                    ... 18 more

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
14 actionable tasks: 4 executed, 10 up-to-date
* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 12s

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for com.lihaoyi:upickle_3:4.1.0-test-publish failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 14s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 4.4.0 at 2026-05-05T01:09:06Z with exit code 1 after 15s
❌ ERROR: 'native-image run' for 4.4.0 failed with exit code 1.
SOFT-FAIL[native-image run][4.4.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.4.0" ./gradlew clean test -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"]
========================================================================================
 com.lihaoyi:upickle_3:4.1.0-test-publish:4.4.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 4.4.0 at 2026-05-05T01:09:06Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.4.0" ./gradlew clean compileTestJava -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 776ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileTestJava NO-SOURCE

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 808ms
9 actionable tasks: 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 4.4.0 at 2026-05-05T01:09:08Z with exit code 0 after 2s
========================================================================================
 com.lihaoyi:upickle_3:4.1.0-test-publish:4.4.0 stage java run
========================================================================================
Starting stage 'java run' for 4.4.0 at 2026-05-05T01:09:08Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.4.0" ./gradlew clean javaTest -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 772ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :compileJava NO-SOURCE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :compileScala NO-SOURCE
> Task :tck-test-support:jar UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE
> Task :compileTestScala
> Task :processTestResources NO-SOURCE
> Task :testClasses

> Task :test FAILED

Upickle_3Test > readsDefaultsAndReportsInvalidInputErrors() FAILED
    org.opentest4j.AssertionFailedError: Unexpected exception type thrown, expected: <upickle.core.AbortException> but was: <upickle.core.TraceVisitor.TraceException>
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:67)
        at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
        at app//org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3083)
        at app//com_lihaoyi.upickle_3.Upickle_3Test.readsDefaultsAndReportsInvalidInputErrors(Upickle_3Test.scala:212)

        Caused by:
        upickle.core.TraceVisitor$TraceException: $

            Caused by:
            upickle.core.AbortException: missing keys in dictionary: age, address, tags, scores, nickname at index 42
                at app//ujson.CharParser$$anon$1.applyOrElse(CharParser.scala:359)
                at app//ujson.CharParser$$anon$1.applyOrElse(CharParser.scala:357)
                at app//scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
                at app//ujson.CharParser.tryCloseCollection(CharParser.scala:512)
                at app//ujson.CharParser.parseNested(CharParser.scala:478)
                at app//ujson.CharParser.parseTopLevel0(CharParser.scala:339)
                at app//ujson.CharParser.parseTopLevel(CharParser.scala:323)

9 tests completed, 1 failed
                at app//ujson.CharParser.parse(CharParser.scala:72)

                at app//ujson.StringParser$.transform(StringParser.scala:28)
                at app//ujson.StringParser$.transform(StringParser.scala:28)
                at app//ujson.Readable$fromTransformer.transform(Readable.scala:19)
                at app//upickle.Api.read$$anonfun$1(Api.scala:50)
                at app//upickle.core.TraceVisitor$.withTrace(TraceVisitor.scala:21)
                at app//upickle.Api.read(Api.scala:50)
                at app//upickle.Api.read$(Api.scala:16)
                at app//upickle.default$.read(Api.scala:187)
                at app//com_lihaoyi.upickle_3.Upickle_3Test.readsDefaultsAndReportsInvalidInputErrors$$anonfun$1(Upickle_3Test.scala:213)
                at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:53)
                at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
                at app//org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3083)
                at app//com_lihaoyi.upickle_3.Upickle_3Test.readsDefaultsAndReportsInvalidInputErrors(Upickle_3Test.scala:212)

                Caused by:
                upickle.core.Abort: missing keys in dictionary: age, address, tags, scores, nickname
                    at app//upickle.implicits.CaseObjectContext.errorMissingKeys(ObjectContexts.scala:43)
                    at app//upickle.implicits.ReadersVersionSpecific$CaseClassReader3V2$ObjectContext.visitEnd(Readers.scala:69)
                    at app//upickle.implicits.ReadersVersionSpecific$CaseClassReader3V2$ObjectContext.visitEnd$(Readers.scala:27)
                    at app//upickle.implicits.ReadersVersionSpecific$$anon$1.visitEnd(Readers.scala:77)
                    at app//upickle.core.TraceVisitor$$anon$1.visitEnd(TraceVisitor.scala:122)
                    ... 18 more

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 12s
13 actionable tasks: 3 executed, 10 up-to-date

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 14s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 4.4.0 at 2026-05-05T01:09:23Z with exit code 1 after 15s
❌ ERROR: 'java run' for 4.4.0 failed with exit code 1.
FAILED[java run][4.4.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.4.0" ./gradlew clean javaTest -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"]
```

## java run on ubuntu-latest / JDK latest-ea / mode current-defaults

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: current-defaults
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.4.0" ./gradlew clean javaTest -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25352097855/job/74333935714
Last 300 lines of the log:
```console

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
14 actionable tasks: 4 executed, 10 up-to-date

9 tests completed, 1 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 13s

> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for com.lihaoyi:upickle_3:4.1.0-test-publish failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 15s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 4.4.0 at 2026-05-05T01:08:31Z with exit code 1 after 15s
❌ ERROR: 'native-image run' for 4.4.0 failed with exit code 1.
SOFT-FAIL[native-image run][4.4.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.4.0" ./gradlew clean test -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"]
========================================================================================
 com.lihaoyi:upickle_3:4.1.0-test-publish:4.4.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 4.4.0 at 2026-05-05T01:08:31Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.4.0" ./gradlew clean compileTestJava -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 738ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :compileScala NO-SOURCE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :compileTestJava NO-SOURCE

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 763ms
9 actionable tasks: 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 4.4.0 at 2026-05-05T01:08:34Z with exit code 0 after 3s
========================================================================================
 com.lihaoyi:upickle_3:4.1.0-test-publish:4.4.0 stage java run
========================================================================================
Starting stage 'java run' for 4.4.0 at 2026-05-05T01:08:34Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.4.0" ./gradlew clean javaTest -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 850ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :processTestResources NO-SOURCE
> Task :compileTestJava NO-SOURCE
> Task :compileTestScala
> Task :testClasses

> Task :test FAILED

Upickle_3Test > readsDefaultsAndReportsInvalidInputErrors() FAILED
    org.opentest4j.AssertionFailedError: Unexpected exception type thrown, expected: <upickle.core.AbortException> but was: <upickle.core.TraceVisitor.TraceException>
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:67)
        at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
        at app//org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3083)
        at app//com_lihaoyi.upickle_3.Upickle_3Test.readsDefaultsAndReportsInvalidInputErrors(Upickle_3Test.scala:212)

        Caused by:
        upickle.core.TraceVisitor$TraceException: $

            Caused by:
            upickle.core.AbortException: missing keys in dictionary: age, address, tags, scores, nickname at index 42
                at app//ujson.CharParser$$anon$1.applyOrElse(CharParser.scala:359)
                at app//ujson.CharParser$$anon$1.applyOrElse(CharParser.scala:357)
                at app//scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
                at app//ujson.CharParser.tryCloseCollection(CharParser.scala:512)
                at app//ujson.CharParser.parseNested(CharParser.scala:478)
                at app//ujson.CharParser.parseTopLevel0(CharParser.scala:339)
                at app//ujson.CharParser.parseTopLevel(CharParser.scala:323)
                at app//ujson.CharParser.parse(CharParser.scala:72)
                at app//ujson.StringParser$.transform(StringParser.scala:28)
                at app//ujson.StringParser$.transform(StringParser.scala:28)
                at app//ujson.Readable$fromTransformer.transform(Readable.scala:19)
                at app//upickle.Api.read$$anonfun$1(Api.scala:50)
                at app//upickle.core.TraceVisitor$.withTrace(TraceVisitor.scala:21)
                at app//upickle.Api.read(Api.scala:50)
                at app//upickle.Api.read$(Api.scala:16)
                at app//upickle.default$.read(Api.scala:187)
                at app//com_lihaoyi.upickle_3.Upickle_3Test.readsDefaultsAndReportsInvalidInputErrors$$anonfun$1(Upickle_3Test.scala:213)
                at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:53)
                at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
                at app//org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3083)
                at app//com_lihaoyi.upickle_3.Upickle_3Test.readsDefaultsAndReportsInvalidInputErrors(Upickle_3Test.scala:212)

                Caused by:
                upickle.core.Abort: missing keys in dictionary: age, address, tags, scores, nickname
                    at app//upickle.implicits.CaseObjectContext.errorMissingKeys(ObjectContexts.scala:43)
                    at app//upickle.implicits.ReadersVersionSpecific$CaseClassReader3V2$ObjectContext.visitEnd(Readers.scala:69)
                    at app//upickle.implicits.ReadersVersionSpecific$CaseClassReader3V2$ObjectContext.visitEnd$(Readers.scala:27)
                    at app//upickle.implicits.ReadersVersionSpecific$$anon$1.visitEnd(Readers.scala:77)
                    at app//upickle.core.TraceVisitor$$anon$1.visitEnd(TraceVisitor.scala:122)
                    ... 18 more

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
13 actionable tasks: 3 executed, 10 up-to-date

9 tests completed, 1 failed

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 12s

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 13s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 4.4.0 at 2026-05-05T01:08:48Z with exit code 1 after 14s
❌ ERROR: 'java run' for 4.4.0 failed with exit code 1.
FAILED[java run][4.4.0][GVM_TCK_NATIVE_IMAGE_MODE="current-defaults" GVM_TCK_LV="4.4.0" ./gradlew clean javaTest -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"]
```

## java run on ubuntu-latest / JDK latest-ea / mode future-defaults-all

Failure kind: java run
OS: ubuntu-latest
JDK: latest-ea
Native-image mode: future-defaults-all
Reproducer: `GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.4.0" ./gradlew clean javaTest -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"`
Runner log: https://github.com/oracle/graalvm-reachability-metadata/actions/runs/25352097855/job/74333935712
Last 300 lines of the log:
```console

9 tests completed, 1 failed

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/tests/test/index.html

* Try:
> Run with 
Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
--scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 13s14 actionable tasks: 4 executed, 10 up-to-date


> Task :tckTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':tckTest'.
> Test for com.lihaoyi:upickle_3:4.1.0-test-publish failed with exit code 1.

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 14s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'native-image run' for 4.4.0 at 2026-05-05T01:09:06Z with exit code 1 after 15s
❌ ERROR: 'native-image run' for 4.4.0 failed with exit code 1.
SOFT-FAIL[native-image run][4.4.0][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.4.0" ./gradlew clean test -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"]
========================================================================================
 com.lihaoyi:upickle_3:4.1.0-test-publish:4.4.0 stage javac compile
========================================================================================
Starting stage 'javac compile' for 4.4.0 at 2026-05-05T01:09:06Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.4.0" ./gradlew clean compileTestJava -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 770ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :compileTestJava
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileTestJava NO-SOURCE

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 794ms
9 actionable tasks: 9 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 2s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'javac compile' for 4.4.0 at 2026-05-05T01:09:09Z with exit code 0 after 3s
========================================================================================
 com.lihaoyi:upickle_3:4.1.0-test-publish:4.4.0 stage java run
========================================================================================
Starting stage 'java run' for 4.4.0 at 2026-05-05T01:09:09Z
Command: GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.4.0" ./gradlew clean javaTest -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE

> Configure project :
GraalVM Reachability Metadata TCK
---------------------------------

> Task :tckClean
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :clean

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 726ms
9 actionable tasks: 1 executed, 8 up-to-date
Consider enabling configuration cache to speed up this build: https://docs.gradle.org/9.1.0/userguide/configuration_cache_enabling.html

> Task :clean

> Task :javaTest
> Task :tck-build-logic:extractPluginRequests UP-TO-DATE
> Task :tck-build-logic:generatePluginAdapters UP-TO-DATE
> Task :tck-build-logic:compileJava UP-TO-DATE
> Task :tck-build-logic:compileGroovy UP-TO-DATE
> Task :tck-build-logic:compileGroovyPlugins UP-TO-DATE
> Task :tck-build-logic:pluginDescriptors UP-TO-DATE
> Task :tck-build-logic:processResources UP-TO-DATE
> Task :tck-build-logic:classes UP-TO-DATE
> Task :tck-build-logic:jar UP-TO-DATE
> Task :checkTestTimeoutAnnotations
> Task :tck-test-support:compileJava UP-TO-DATE
> Task :compileJava NO-SOURCE
> Task :tck-test-support:processResources NO-SOURCE
> Task :tck-test-support:classes UP-TO-DATE
> Task :compileScala NO-SOURCE
> Task :processResources NO-SOURCE
> Task :classes UP-TO-DATE
> Task :tck-test-support:jar UP-TO-DATE
> Task :compileTestJava NO-SOURCE
> Task :compileTestScala
> Task :processTestResources NO-SOURCE
> Task :testClasses

> Task :test FAILED

Upickle_3Test > readsDefaultsAndReportsInvalidInputErrors() FAILED
    org.opentest4j.AssertionFailedError: Unexpected exception type thrown, expected: <upickle.core.AbortException> but was: <upickle.core.TraceVisitor.TraceException>
        at app//org.junit.jupiter.api.AssertionFailureBuilder.build(AssertionFailureBuilder.java:151)
        at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:67)
        at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
        at app//org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3083)
        at app//com_lihaoyi.upickle_3.Upickle_3Test.readsDefaultsAndReportsInvalidInputErrors(Upickle_3Test.scala:212)

        Caused by:
        upickle.core.TraceVisitor$TraceException: $

            Caused by:
            upickle.core.AbortException: missing keys in dictionary: age, address, tags, scores, nickname at index 42
                at app//ujson.CharParser$$anon$1.applyOrElse(CharParser.scala:359)
                at app//ujson.CharParser$$anon$1.applyOrElse(CharParser.scala:357)
                at app//scala.runtime.AbstractPartialFunction.apply(AbstractPartialFunction.scala:35)
                at app//ujson.CharParser.tryCloseCollection(CharParser.scala:512)
                at app//ujson.CharParser.parseNested(CharParser.scala:478)
                at app//ujson.CharParser.parseTopLevel0(CharParser.scala:339)
                at app//ujson.CharParser.parseTopLevel(CharParser.scala:323)
                at app//ujson.CharParser.parse(CharParser.scala:72)
                at app//ujson.StringParser$.transform(StringParser.scala:28)
                at app//ujson.StringParser$.transform(StringParser.scala:28)
                at app//ujson.Readable$fromTransformer.transform(Readable.scala:19)
                at app//upickle.Api.read$$anonfun$1(Api.scala:50)
                at app//upickle.core.TraceVisitor$.withTrace(TraceVisitor.scala:21)
                at app//upickle.Api.read(Api.scala:50)
                at app//upickle.Api.read$(Api.scala:16)
                at app//upickle.default$.read(Api.scala:187)
                at app//com_lihaoyi.upickle_3.Upickle_3Test.readsDefaultsAndReportsInvalidInputErrors$$anonfun$1(Upickle_3Test.scala:213)
                at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:53)
                at app//org.junit.jupiter.api.AssertThrows.assertThrows(AssertThrows.java:35)
                at app//org.junit.jupiter.api.Assertions.assertThrows(Assertions.java:3083)
                at app//com_lihaoyi.upickle_3.Upickle_3Test.readsDefaultsAndReportsInvalidInputErrors(Upickle_3Test.scala:212)

                Caused by:
                upickle.core.Abort: missing keys in dictionary: age, address, tags, scores, nickname
                    at app//upickle.implicits.CaseObjectContext.errorMissingKeys(ObjectContexts.scala:43)
                    at app//upickle.implicits.ReadersVersionSpecific$CaseClassReader3V2$ObjectContext.visitEnd(Readers.scala:69)
                    at app//upickle.implicits.ReadersVersionSpecific$CaseClassReader3V2$ObjectContext.visitEnd$(Readers.scala:27)
                    at app//upickle.implicits.ReadersVersionSpecific$$anon$1.visitEnd(Readers.scala:77)
                    at app//upickle.core.TraceVisitor$$anon$1.visitEnd(TraceVisitor.scala:122)
                    ... 18 more

9 tests completed, 1 failed

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/problems/problems-report.html

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.


FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':test'.
> There were failing tests. See the report at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/tests/src/com.lihaoyi/upickle_3/4.1.0-test-publish/build/reports/tests/test/index.html

* Try:
> Run with --scan to generate a Build Scan (Powered by Develocity).

BUILD FAILED in 12s
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.
13 actionable tasks: 3 executed, 10 up-to-date

> Task :javaTest FAILED

[Incubating] Problems report is available at: file:///home/runner/work/graalvm-reachability-metadata/graalvm-reachability-metadata/build/reports/problems/problems-report.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':javaTest'.
> Java tests failed

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to generate a Build Scan (Powered by Develocity).
> Get more help at https://help.gradle.org.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.1.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD FAILED in 13s
11 actionable tasks: 3 executed, 8 up-to-date
Finished stage 'java run' for 4.4.0 at 2026-05-05T01:09:23Z with exit code 1 after 14s
❌ ERROR: 'java run' for 4.4.0 failed with exit code 1.
FAILED[java run][4.4.0][GVM_TCK_NATIVE_IMAGE_MODE="future-defaults-all" GVM_TCK_LV="4.4.0" ./gradlew clean javaTest -Pcoordinates="com.lihaoyi:upickle_3:4.1.0-test-publish"]
```

--- a/metadata/com.lihaoyi/upickle_3/4.4.0/reachability-metadata.json
+++ b/metadata/com.lihaoyi/upickle_3/4.4.0/reachability-metadata.json
@@ -1,0 +1,82 @@
+{
+  "reflection" : [
+    {
+      "condition" : {
+        "typeReached" : "upickle.default$"
+      },
+      "type" : "java.lang.ClassValue"
+    },
+    {
+      "condition" : {
+        "typeReached" : "upickle.default$"
+      },
+      "type" : "java.lang.invoke.VarHandle",
+      "methods" : [
+        {
+          "name" : "releaseFence",
+          "parameterTypes" : [ ]
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "upickle.default$"
+      },
+      "type" : "sun.misc.Unsafe",
+      "fields" : [
+        {
+          "name" : "theUnsafe"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "upickle.default$"
+      },
+      "type" : "upickle.default$",
+      "fields" : [
+        {
+          "name" : "ReadWriter$lzy1"
+        },
+        {
+          "name" : "Reader$lzy1"
+        },
+        {
+          "name" : "TaggedReadWriter$lzy1"
+        },
+        {
+          "name" : "TaggedReader$lzy1"
+        },
+        {
+          "name" : "TaggedWriter$lzy1"
+        },
+        {
+          "name" : "Writer$lzy1"
+        },
+        {
+          "name" : "transform$lzy1"
+        },
+        {
+          "name" : "upickle$implicits$Readers$$JavaReader$lzy1"
+        },
+        {
+          "name" : "upickle$implicits$Writers$$JavaWriter$lzy1"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "upickle.implicits.ReadersVersionSpecific$CaseClassReader3V2"
+      },
+      "type" : "upickle.implicits.ReadersVersionSpecific$CaseClassReader3V2",
+      "fields" : [
+        {
+          "name" : "$1$$lzy1"
+        },
+        {
+          "name" : "hasFlattenOnMap$lzy1"
+        }
+      ]
+    }
+  ]
+}

--- a/metadata/com.lihaoyi/upickle_3/index.json
+++ b/metadata/com.lihaoyi/upickle_3/index.json
@@ -2,6 +2,11 @@
   {
     "latest" : true,
     "metadata-version" : "4.4.0",
+    "source-code-url" : "https://repo.maven.apache.org/maven2/com/lihaoyi/upickle_3/$version$/upickle_3-$version$-sources.jar",
+    "repository-url" : "https://github.com/com-lihaoyi/upickle",
+    "test-code-url" : "https://github.com/com-lihaoyi/upickle/tree/$version$/upickle/test/src",
+    "documentation-url" : "https://repo.maven.apache.org/maven2/com/lihaoyi/upickle_3/$version$/upickle_3-$version$-javadoc.jar",
+    "description" : "uPickle is a Scala JSON and MessagePack serialization library for reading and writing structured data. The upickle_3 artifact provides the Scala 3 JVM API that combines JSON and MessagePack readers, writers, macro derivation, and default serialization helpers for application code.",
     "language" : {
       "name" : "scala",
       "version" : "3"

--- a/metadata/com.lihaoyi/upickle_3/index.json
+++ b/metadata/com.lihaoyi/upickle_3/index.json
@@ -1,6 +1,19 @@
 [
   {
     "latest" : true,
+    "metadata-version" : "4.4.0",
+    "language" : {
+      "name" : "scala",
+      "version" : "3"
+    },
+    "tested-versions" : [
+      "4.4.0"
+    ],
+    "allowed-packages" : [
+      "upickle"
+    ]
+  },
+  {
     "metadata-version" : "4.1.0-test-publish",
     "source-code-url" : "https://repo.maven.apache.org/maven2/com/lihaoyi/upickle_3/$version$/upickle_3-$version$-sources.jar",
     "repository-url" : "https://github.com/com-lihaoyi/upickle",

--- a/stats/com.lihaoyi/upickle_3/4.4.0/execution-metrics.json
+++ b/stats/com.lihaoyi/upickle_3/4.4.0/execution-metrics.json
@@ -1,0 +1,88 @@
+{
+  "fix_java_run_fail:2026-05-05": {
+    "timestamp": "2026-05-05T03:15:23.832944Z",
+    "library": "com.lihaoyi:upickle_3:4.4.0",
+    "starting_commit": "8caac2d8b97650469662ebc22a62ba97e3894a80",
+    "ending_commit": "fbc16e1cf0e926912077296a06a0da7856bd68e0",
+    "previous_library": "com.lihaoyi:upickle_3:4.1.0-test-publish",
+    "strategy_name": "java_run_iterative_with_coverage_sources_pi_gpt-5.5",
+    "agent": "pi",
+    "model": "oca/gpt-5.5",
+    "status": "success",
+    "stats": {
+      "version": "4.4.0",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 1253,
+          "missed": 8589,
+          "ratio": 0.127312,
+          "total": 9842
+        },
+        "line": {
+          "covered": 70,
+          "missed": 103,
+          "ratio": 0.404624,
+          "total": 173
+        },
+        "method": {
+          "covered": 160,
+          "missed": 1191,
+          "ratio": 0.118431,
+          "total": 1351
+        }
+      }
+    },
+    "previous_library_stats": {
+      "version": "4.1.0-test-publish",
+      "dynamicAccess": {
+        "breakdown": {},
+        "coverageRatio": 1.0,
+        "coveredCalls": 0,
+        "totalCalls": 0
+      },
+      "libraryCoverage": {
+        "instruction": {
+          "covered": 1277,
+          "missed": 5985,
+          "ratio": 0.175847,
+          "total": 7262
+        },
+        "line": {
+          "covered": 72,
+          "missed": 100,
+          "ratio": 0.418605,
+          "total": 172
+        },
+        "method": {
+          "covered": 160,
+          "missed": 765,
+          "ratio": 0.172973,
+          "total": 925
+        }
+      }
+    },
+    "metrics": {
+      "input_tokens_used": 138839,
+      "cached_input_tokens_used": 2312192,
+      "output_tokens_used": 13055,
+      "iterations": 2,
+      "cost_usd": 1.5477,
+      "tested_library_loc": 70,
+      "metadata_entries": 18,
+      "test_only_metadata_entries": 19,
+      "previous_library_metadata_entries": 0,
+      "code_coverage_percent": 40.46,
+      "previous_library_coverage_percent": 41.86
+    },
+    "artifacts": {
+      "test_file": "tests/src/com.lihaoyi/upickle_3/4.4.0/src/test/scala/com_lihaoyi/upickle_3/Upickle_3Test.scala",
+      "metadata_file": "metadata/com.lihaoyi/upickle_3/4.4.0/reachability-metadata.json"
+    }
+  }
+}

--- a/stats/com.lihaoyi/upickle_3/4.4.0/stats.json
+++ b/stats/com.lihaoyi/upickle_3/4.4.0/stats.json
@@ -1,0 +1,31 @@
+{
+  "versions" : [ {
+    "version" : "4.4.0",
+    "dynamicAccess" : {
+      "breakdown" : { },
+      "coverageRatio" : 1.0,
+      "coveredCalls" : 0,
+      "totalCalls" : 0
+    },
+    "libraryCoverage" : {
+      "instruction" : {
+        "covered" : 1253,
+        "missed" : 8589,
+        "ratio" : 0.127312,
+        "total" : 9842
+      },
+      "line" : {
+        "covered" : 70,
+        "missed" : 103,
+        "ratio" : 0.404624,
+        "total" : 173
+      },
+      "method" : {
+        "covered" : 160,
+        "missed" : 1191,
+        "ratio" : 0.118431,
+        "total" : 1351
+      }
+    }
+  } ]
+}

--- a/tests/src/com.lihaoyi/upickle_3/4.4.0/.gitignore
+++ b/tests/src/com.lihaoyi/upickle_3/4.4.0/.gitignore
@@ -1,0 +1,4 @@
+gradlew.bat
+gradlew
+gradle/
+build/

--- a/tests/src/com.lihaoyi/upickle_3/4.4.0/build.gradle
+++ b/tests/src/com.lihaoyi/upickle_3/4.4.0/build.gradle
@@ -1,0 +1,36 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+plugins {
+    id "org.graalvm.internal.tck"
+    id "scala"
+}
+
+String libraryVersion = tck.testedLibraryVersion.get()
+
+dependencies {
+    testImplementation "com.lihaoyi:upickle_3:$libraryVersion"
+    testImplementation 'org.assertj:assertj-core:3.22.0'
+    testImplementation 'org.scala-lang:scala3-library_3:3.3.6'
+    testImplementation 'org.scala-lang:scala3-compiler_3:3.3.6'
+}
+
+java {
+    toolchain {
+        languageVersion = JavaLanguageVersion.of(21)
+    }
+}
+
+graalvmNative {
+    agent {
+        defaultMode = "conditional"
+        modes {
+            conditional {
+                userCodeFilterPath = "user-code-filter.json"
+            }
+        }
+    }
+}

--- a/tests/src/com.lihaoyi/upickle_3/4.4.0/build.gradle
+++ b/tests/src/com.lihaoyi/upickle_3/4.4.0/build.gradle
@@ -4,6 +4,16 @@
  * You should have received a copy of the CC0 legalcode along with this
  * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
  */
+import org.gradle.api.tasks.testing.Test
+
+import java.io.File
+
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
 plugins {
     id "org.graalvm.internal.tck"
     id "scala"
@@ -16,6 +26,19 @@ dependencies {
     testImplementation 'org.assertj:assertj-core:3.22.0'
     testImplementation 'org.scala-lang:scala3-library_3:3.3.6'
     testImplementation 'org.scala-lang:scala3-compiler_3:3.3.6'
+}
+
+String graalvmHome = System.getenv("GRAALVM_HOME") ?: System.getenv("JAVA_HOME")
+File graalvmJavaExecutable = graalvmHome == null ? null : new File(graalvmHome, "bin/java")
+File graalvmLibDirectory = graalvmHome == null ? null : new File(graalvmHome, "lib")
+
+if (project.hasProperty("agent") && graalvmJavaExecutable != null && graalvmJavaExecutable.isFile()) {
+    tasks.withType(Test).configureEach {
+        executable = graalvmJavaExecutable.absolutePath
+        if (graalvmLibDirectory != null && graalvmLibDirectory.isDirectory()) {
+            jvmArgs("-Djava.library.path=${graalvmLibDirectory.absolutePath}")
+        }
+    }
 }
 
 java {

--- a/tests/src/com.lihaoyi/upickle_3/4.4.0/gradle.properties
+++ b/tests/src/com.lihaoyi/upickle_3/4.4.0/gradle.properties
@@ -1,0 +1,6 @@
+# Optional explicit tested library coordinates (format: group:artifact:version).
+# If omitted, fallback resolution in org.graalvm.internal.tck.gradle still uses
+# environment variable GVM_TCK_LC and then this property.
+library.coordinates = com.lihaoyi:upickle_3:4.4.0
+library.version = 4.4.0
+metadata.dir = com.lihaoyi/upickle_3/4.4.0/

--- a/tests/src/com.lihaoyi/upickle_3/4.4.0/settings.gradle
+++ b/tests/src/com.lihaoyi/upickle_3/4.4.0/settings.gradle
@@ -1,0 +1,20 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+pluginManagement {
+    def tckPath = Objects.requireNonNullElse(
+            System.getenv("GVM_TCK_TCKDIR"),
+            "../../../../tck-build-logic"
+    )
+    includeBuild(tckPath)
+    repositories { mavenLocal(); gradlePluginPortal() }
+}
+
+plugins {
+    id "org.graalvm.internal.tck-settings" version "1.0.0-SNAPSHOT"
+}
+
+rootProject.name = 'com.lihaoyi.upickle_3_tests'

--- a/tests/src/com.lihaoyi/upickle_3/4.4.0/src/test/resources/META-INF/native-image/reachability-metadata.json
+++ b/tests/src/com.lihaoyi/upickle_3/4.4.0/src/test/resources/META-INF/native-image/reachability-metadata.json
@@ -1,0 +1,109 @@
+{
+  "reflection" : [
+    {
+      "condition" : {
+        "typeReached" : "com_lihaoyi.upickle_3.Address"
+      },
+      "type" : "com_lihaoyi.upickle_3.Address$",
+      "fields" : [
+        {
+          "name" : "derived$ReadWriter$lzy1"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "com_lihaoyi.upickle_3.Customer"
+      },
+      "type" : "com_lihaoyi.upickle_3.Customer$",
+      "fields" : [
+        {
+          "name" : "derived$ReadWriter$lzy2"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "com_lihaoyi.upickle_3.DefaultsEnabled"
+      },
+      "type" : "com_lihaoyi.upickle_3.DefaultsEnabled$",
+      "fields" : [
+        {
+          "name" : "derived$ReadWriter$lzy7"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "com_lihaoyi.upickle_3.Email"
+      },
+      "type" : "com_lihaoyi.upickle_3.Email$",
+      "fields" : [
+        {
+          "name" : "given_ReadWriter_Email$lzy1"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "com_lihaoyi.upickle_3.InventoryEvent"
+      },
+      "type" : "com_lihaoyi.upickle_3.InventoryEvent$",
+      "fields" : [
+        {
+          "name" : "derived$ReadWriter$lzy6"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "com_lihaoyi.upickle_3.InventoryEvent$Added"
+      },
+      "type" : "com_lihaoyi.upickle_3.InventoryEvent$Added$",
+      "fields" : [
+        {
+          "name" : "derived$ReadWriter$lzy3"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "com_lihaoyi.upickle_3.InventoryEvent$Recounted"
+      },
+      "type" : "com_lihaoyi.upickle_3.InventoryEvent$Recounted$",
+      "fields" : [
+        {
+          "name" : "derived$ReadWriter$lzy5"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "com_lihaoyi.upickle_3.InventoryEvent$Removed"
+      },
+      "type" : "com_lihaoyi.upickle_3.InventoryEvent$Removed$",
+      "fields" : [
+        {
+          "name" : "derived$ReadWriter$lzy4"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "com_lihaoyi.upickle_3.ProductCode"
+      },
+      "type" : "com_lihaoyi.upickle_3.ProductCode$",
+      "fields" : [
+        {
+          "name" : "given_ReadWriter_ProductCode$lzy1"
+        }
+      ]
+    },
+    {
+      "condition" : {
+        "typeReached" : "com_lihaoyi.upickle_3.Customer$$anon$3"
+      },
+      "type" : "java.lang.String[]"
+    }
+  ]
+}

--- a/tests/src/com.lihaoyi/upickle_3/4.4.0/src/test/scala/com_lihaoyi/upickle_3/Upickle_3Test.scala
+++ b/tests/src/com.lihaoyi/upickle_3/4.4.0/src/test/scala/com_lihaoyi/upickle_3/Upickle_3Test.scala
@@ -1,0 +1,307 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package com_lihaoyi.upickle_3
+
+import org.junit.jupiter.api.Assertions.assertArrayEquals
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertFalse
+import org.junit.jupiter.api.Assertions.assertNotEquals
+import org.junit.jupiter.api.Assertions.assertThrows
+import org.junit.jupiter.api.Assertions.assertTrue
+import org.junit.jupiter.api.Test
+import upickle.default.*
+import upickle.implicits.key
+
+import java.io.ByteArrayOutputStream
+import java.io.StringWriter
+import java.nio.charset.StandardCharsets
+import java.util.Locale
+
+final case class Address(street: String, zipCode: Int) derives ReadWriter
+
+final case class Customer(
+    @key("customer_name") name: String,
+    age: Int,
+    address: Address,
+    tags: Vector[String],
+    scores: Map[String, Double],
+    nickname: Option[String]
+) derives ReadWriter
+
+sealed trait InventoryEvent derives ReadWriter
+
+object InventoryEvent {
+  final case class Added(sku: String, quantity: Int, customer: Customer) extends InventoryEvent derives ReadWriter
+  final case class Removed(sku: String, reason: Option[String]) extends InventoryEvent derives ReadWriter
+  final case class Recounted(sku: String, previous: Int, current: Int) extends InventoryEvent derives ReadWriter
+}
+
+final case class Email(value: String)
+
+object Email {
+  given ReadWriter[Email] = readwriter[String].bimap[Email](
+    email => email.value,
+    value => Email(value.trim.toLowerCase(Locale.ROOT))
+  )
+}
+
+final case class Contact(name: String, primaryEmail: Email, secondaryEmails: List[Email]) derives ReadWriter
+
+final case class ProductCode(prefix: String, number: Int)
+
+object ProductCode {
+  given ReadWriter[ProductCode] = stringKeyRW(readwriter[String].bimap[ProductCode](
+    code => s"${code.prefix}-${code.number}",
+    value => {
+      val parts: Array[String] = value.split("-", 2)
+      ProductCode(parts(0), parts(1).toInt)
+    }
+  ))
+}
+
+final case class DefaultsEnabled(name: String, active: Boolean = true, aliases: List[String] = Nil) derives ReadWriter
+
+final case class ExternalLine(sku: String, quantity: Int) derives ReadWriter
+
+final case class ExternalOrder(id: String, lines: List[ExternalLine], metadata: Map[String, String]) derives ReadWriter
+
+final case class InternalLine(sku: String, quantity: Int) derives ReadWriter
+
+final case class InternalOrder(id: String, lines: List[InternalLine], metadata: Map[String, String]) derives ReadWriter
+
+class Upickle_3Test {
+  @Test
+  def readsAndWritesNestedCaseClassesWithCollectionsOptionsAndRenamedFields(): Unit = {
+    val customer: Customer = Customer(
+      name = "Ada Lovelace",
+      age = 36,
+      address = Address("St James's Square", 101),
+      tags = Vector("founder", "math", "unicode-☃"),
+      scores = Map("analysis" -> 99.5, "poetry" -> 88.25),
+      nickname = Some("enchantress")
+    )
+
+    val jsonText: String = write(customer, 2, false, true)
+    val jsonValue: ujson.Value = ujson.read(jsonText)
+
+    assertEquals("Ada Lovelace", jsonValue(objKey("customer_name")).str)
+    assertFalse(jsonValue.obj.contains("name"))
+    assertEquals(36.0, jsonValue(objKey("age")).num, 0.0)
+    assertEquals("St James's Square", jsonValue(objKey("address"))(objKey("street")).str)
+    assertEquals(101.0, jsonValue(objKey("address"))(objKey("zipCode")).num, 0.0)
+    assertEquals(List("founder", "math", "unicode-☃"), jsonValue(objKey("tags")).arr.map(_.str).toList)
+    assertEquals("enchantress", jsonValue(objKey("nickname")).str)
+
+    val parsed: Customer = read[Customer](jsonText)
+    assertEquals(customer, parsed)
+
+    val withoutNickname: Customer = customer.copy(nickname = None)
+    val withoutNicknameJson: ujson.Value = writeJs(withoutNickname)
+    assertTrue(withoutNicknameJson(objKey("nickname")).isNull)
+    assertEquals(withoutNickname, read[Customer](withoutNicknameJson))
+  }
+
+  @Test
+  def roundTripsSealedTraitHierarchiesThroughJsonAndMessagePack(): Unit = {
+    val customer: Customer = Customer(
+      "Grace Hopper",
+      85,
+      Address("Arlington", 42),
+      Vector("compiler", "navy"),
+      Map("debugging" -> 100.0),
+      None
+    )
+    val events: List[InventoryEvent] = List(
+      InventoryEvent.Added("BOOK-1", 4, customer),
+      InventoryEvent.Removed("BOOK-2", Some("damaged")),
+      InventoryEvent.Recounted("BOOK-3", previous = 7, current = 9)
+    )
+
+    val jsonText: String = write(events)
+    assertEquals(events, read[List[InventoryEvent]](jsonText))
+
+    val jsonAst: ujson.Value = writeJs(events)
+    assertTrue(jsonAst.arr.nonEmpty)
+    assertEquals(events, read[List[InventoryEvent]](jsonAst))
+
+    val binary: Array[Byte] = writeBinaryToByteArray(events)
+    assertTrue(binary.length > 0)
+    assertEquals(events, readBinary[List[InventoryEvent]](upack.Readable.fromByteArray(binary)))
+  }
+
+  @Test
+  def supportsPrimitiveCollectionsTuplesAndMaps(): Unit = {
+    val payload: Map[String, List[Option[Int]]] = Map(
+      "present" -> List(Some(1), Some(2), Some(3)),
+      "mixed" -> List(Some(4), None, Some(6)),
+      "empty" -> Nil
+    )
+
+    val jsonText: String = write(payload, -1, false, true)
+    assertEquals(payload, read[Map[String, List[Option[Int]]]](jsonText))
+
+    val tuple: (String, Int, Boolean, Option[Double]) = ("tuple", 7, true, Some(2.5d))
+    assertEquals(tuple, read[(String, Int, Boolean, Option[Double])](write(tuple)))
+
+    val numericKeyedMap: Map[Int, String] = Map(1 -> "one", 2 -> "two", 10 -> "ten")
+    val numericKeyedJson: ujson.Value = writeJs(numericKeyedMap)
+    assertEquals("one", numericKeyedJson(objKey("1")).str)
+    assertEquals(numericKeyedMap, read[Map[Int, String]](numericKeyedJson))
+
+    val tupleValues: List[(String, Int, Boolean)] = List(("one", 1, true), ("two", 2, false))
+    assertEquals(tupleValues, read[List[(String, Int, Boolean)]](write(tupleValues)))
+  }
+
+  @Test
+  def usesCustomReadWritersForValueClassesAndAppliesReadMapping(): Unit = {
+    val contact: Contact = Contact(
+      name = "Support",
+      primaryEmail = Email("Help@Example.COM"),
+      secondaryEmails = List(Email("Admin@Example.COM"), Email("Ops@Example.COM"))
+    )
+
+    val jsonText: String = write(contact)
+    val jsonValue: ujson.Value = ujson.read(jsonText)
+    assertEquals("Help@Example.COM", jsonValue(objKey("primaryEmail")).str)
+    assertEquals(List("Admin@Example.COM", "Ops@Example.COM"), jsonValue(objKey("secondaryEmails")).arr.map(_.str).toList)
+
+    val parsed: Contact = read[Contact](
+      """
+        |{
+        |  "name": "Support",
+        |  "primaryEmail": " HELP@EXAMPLE.COM ",
+        |  "secondaryEmails": [" ADMIN@EXAMPLE.COM ", " OPS@EXAMPLE.COM "]
+        |}
+        |""".stripMargin
+    )
+    assertEquals(Contact("Support", Email("help@example.com"), List(Email("admin@example.com"), Email("ops@example.com"))), parsed)
+  }
+
+  @Test
+  def supportsCustomTypesAsJsonObjectKeysWithStringKeyReadWriter(): Unit = {
+    val stockByProduct: Map[ProductCode, Int] = Map(
+      ProductCode("BOOK", 1) -> 12,
+      ProductCode("PEN", 9) -> 5
+    )
+
+    val jsonValue: ujson.Value = writeJs(stockByProduct)
+    assertEquals(12.0, jsonValue(objKey("BOOK-1")).num, 0.0)
+    assertEquals(5.0, jsonValue(objKey("PEN-9")).num, 0.0)
+    assertEquals(stockByProduct, read[Map[ProductCode, Int]](jsonValue))
+
+    val jsonText: String = write(stockByProduct, -1, false, true)
+    assertTrue(jsonText.startsWith("{"))
+    assertFalse(jsonText.startsWith("["))
+    assertEquals(stockByProduct, read[Map[ProductCode, Int]](jsonText))
+  }
+
+  @Test
+  def readsDefaultsAndReportsInvalidInputErrors(): Unit = {
+    val defaults: DefaultsEnabled = read[DefaultsEnabled]("""{"name":"defaults"}""")
+    assertEquals(DefaultsEnabled("defaults"), defaults)
+
+    val explicit: DefaultsEnabled = DefaultsEnabled("custom", active = false, aliases = List("x", "y"))
+    assertEquals(explicit, read[DefaultsEnabled](write(explicit)))
+
+    assertThrows(
+      classOf[upickle.core.AbortException],
+      () => {
+        read[Customer]("""{"customer_name":"missing-required-fields"}""")
+        ()
+      }
+    )
+
+    assertThrows(
+      classOf[upickle.core.AbortException],
+      () => {
+        read[List[Int]]("""{"not":"a-list"}""")
+        ()
+      }
+    )
+  }
+
+  @Test
+  def transformsBetweenCompatibleTypesUsingReadersAndWriters(): Unit = {
+    val external: ExternalOrder = ExternalOrder(
+      id = "order-42",
+      lines = List(ExternalLine("BOOK-1", 2), ExternalLine("PEN-9", 5)),
+      metadata = Map("channel" -> "web", "priority" -> "standard")
+    )
+
+    val internal: InternalOrder = transform(external).to[InternalOrder]
+
+    assertEquals(
+      InternalOrder(
+        id = "order-42",
+        lines = List(InternalLine("BOOK-1", 2), InternalLine("PEN-9", 5)),
+        metadata = Map("channel" -> "web", "priority" -> "standard")
+      ),
+      internal
+    )
+  }
+
+  @Test
+  def writesToTextAndByteTargetsWithConsistentJsonContent(): Unit = {
+    val customer: Customer = Customer(
+      "Katherine Johnson",
+      101,
+      Address("White Sulphur Springs", 314),
+      Vector("space", "math"),
+      Map("orbital" -> 100.0, "navigation" -> 99.0),
+      Some("computer")
+    )
+
+    val stringWriter: StringWriter = new StringWriter()
+    writeTo(customer, stringWriter, 2, false, true)
+    val fromWriter: String = stringWriter.toString
+    assertTrue(fromWriter.contains("\n"))
+    assertEquals(customer, read[Customer](fromWriter))
+
+    val jsonBytes: Array[Byte] = writeToByteArray(customer, -1, false, true)
+    val fromBytes: String = new String(jsonBytes, StandardCharsets.UTF_8)
+    assertFalse(fromBytes.contains("\n"))
+    assertEquals(customer, read[Customer](ujson.Readable.fromByteArray(jsonBytes)))
+
+    val outputStream: ByteArrayOutputStream = new ByteArrayOutputStream()
+    writeToOutputStream(customer, outputStream, -1, false, true)
+    assertArrayEquals(jsonBytes, outputStream.toByteArray)
+  }
+
+  @Test
+  def escapesUnicodeAndSortsKeysWhenRequested(): Unit = {
+    val customer: Customer = Customer(
+      "Zoë ☃",
+      28,
+      Address("München", 80331),
+      Vector("λ", "雪"),
+      Map("zeta" -> 1.0, "alpha" -> 2.0),
+      None
+    )
+
+    val unescaped: String = write(customer, -1, false, false)
+    val escapedSorted: String = write(customer, -1, true, true)
+
+    assertTrue(unescaped.contains("Zoë ☃"))
+    assertTrue(unescaped.contains("München"))
+    val lowerCaseEscapedSorted: String = escapedSorted.toLowerCase(Locale.ROOT)
+    assertTrue(lowerCaseEscapedSorted.contains("zo\\u00eb \\u2603"))
+    assertTrue(lowerCaseEscapedSorted.contains("m\\u00fcnchen"))
+    assertNotEquals(unescaped, escapedSorted)
+    assertEquals(customer, read[Customer](escapedSorted))
+
+    assertFieldOrder(escapedSorted, Seq("address", "age", "customer_name", "nickname", "scores", "tags"))
+    assertFieldOrder(escapedSorted, Seq("alpha", "zeta"))
+  }
+
+  private def objKey(value: String): ujson.Value.Selector = ujson.Value.Selector.StringSelector(value)
+
+  private def assertFieldOrder(jsonText: String, fields: Seq[String]): Unit = {
+    val positions: Seq[Int] = fields.map(field => jsonText.indexOf(s"\"$field\""))
+    assertTrue(positions.forall(_ >= 0), s"Expected all fields $fields to appear in $jsonText")
+    assertEquals(positions.sorted, positions, s"Expected fields $fields to appear in order in $jsonText")
+  }
+}

--- a/tests/src/com.lihaoyi/upickle_3/4.4.0/src/test/scala/com_lihaoyi/upickle_3/Upickle_3Test.scala
+++ b/tests/src/com.lihaoyi/upickle_3/4.4.0/src/test/scala/com_lihaoyi/upickle_3/Upickle_3Test.scala
@@ -49,8 +49,6 @@ object Email {
   )
 }
 
-final case class Contact(name: String, primaryEmail: Email, secondaryEmails: List[Email]) derives ReadWriter
-
 final case class ProductCode(prefix: String, number: Int)
 
 object ProductCode {
@@ -64,14 +62,6 @@ object ProductCode {
 }
 
 final case class DefaultsEnabled(name: String, active: Boolean = true, aliases: List[String] = Nil) derives ReadWriter
-
-final case class ExternalLine(sku: String, quantity: Int) derives ReadWriter
-
-final case class ExternalOrder(id: String, lines: List[ExternalLine], metadata: Map[String, String]) derives ReadWriter
-
-final case class InternalLine(sku: String, quantity: Int) derives ReadWriter
-
-final case class InternalOrder(id: String, lines: List[InternalLine], metadata: Map[String, String]) derives ReadWriter
 
 class Upickle_3Test {
   @Test
@@ -128,7 +118,7 @@ class Upickle_3Test {
     assertTrue(jsonAst.arr.nonEmpty)
     assertEquals(events, read[List[InventoryEvent]](jsonAst))
 
-    val binary: Array[Byte] = writeBinaryToByteArray(events)
+    val binary: Array[Byte] = writeBinary(events)
     assertTrue(binary.length > 0)
     assertEquals(events, readBinary[List[InventoryEvent]](upack.Readable.fromByteArray(binary)))
   }
@@ -158,27 +148,14 @@ class Upickle_3Test {
 
   @Test
   def usesCustomReadWritersForValueClassesAndAppliesReadMapping(): Unit = {
-    val contact: Contact = Contact(
-      name = "Support",
-      primaryEmail = Email("Help@Example.COM"),
-      secondaryEmails = List(Email("Admin@Example.COM"), Email("Ops@Example.COM"))
-    )
+    val emails: List[Email] = List(Email("Help@Example.COM"), Email("Admin@Example.COM"))
 
-    val jsonText: String = write(contact)
+    val jsonText: String = write(emails)
     val jsonValue: ujson.Value = ujson.read(jsonText)
-    assertEquals("Help@Example.COM", jsonValue(objKey("primaryEmail")).str)
-    assertEquals(List("Admin@Example.COM", "Ops@Example.COM"), jsonValue(objKey("secondaryEmails")).arr.map(_.str).toList)
+    assertEquals(List("Help@Example.COM", "Admin@Example.COM"), jsonValue.arr.map(_.str).toList)
 
-    val parsed: Contact = read[Contact](
-      """
-        |{
-        |  "name": "Support",
-        |  "primaryEmail": " HELP@EXAMPLE.COM ",
-        |  "secondaryEmails": [" ADMIN@EXAMPLE.COM ", " OPS@EXAMPLE.COM "]
-        |}
-        |""".stripMargin
-    )
-    assertEquals(Contact("Support", Email("help@example.com"), List(Email("admin@example.com"), Email("ops@example.com"))), parsed)
+    val parsed: List[Email] = read[List[Email]]("""[" HELP@EXAMPLE.COM ", " ADMIN@EXAMPLE.COM "]""")
+    assertEquals(List(Email("help@example.com"), Email("admin@example.com")), parsed)
   }
 
   @Test
@@ -208,7 +185,7 @@ class Upickle_3Test {
     assertEquals(explicit, read[DefaultsEnabled](write(explicit)))
 
     assertThrows(
-      classOf[upickle.core.AbortException],
+      classOf[upickle.core.TraceVisitor.TraceException],
       () => {
         read[Customer]("""{"customer_name":"missing-required-fields"}""")
         ()
@@ -216,7 +193,7 @@ class Upickle_3Test {
     )
 
     assertThrows(
-      classOf[upickle.core.AbortException],
+      classOf[upickle.core.TraceVisitor.TraceException],
       () => {
         read[List[Int]]("""{"not":"a-list"}""")
         ()
@@ -225,23 +202,19 @@ class Upickle_3Test {
   }
 
   @Test
-  def transformsBetweenCompatibleTypesUsingReadersAndWriters(): Unit = {
-    val external: ExternalOrder = ExternalOrder(
-      id = "order-42",
-      lines = List(ExternalLine("BOOK-1", 2), ExternalLine("PEN-9", 5)),
-      metadata = Map("channel" -> "web", "priority" -> "standard")
+  def transformsBetweenCompatibleReadersAndWriters(): Unit = {
+    val customer: Customer = Customer(
+      "Dorothy Vaughan",
+      98,
+      Address("Hampton", 1),
+      Vector("manager", "fortran"),
+      Map("leadership" -> 100.0),
+      Some("human computer")
     )
 
-    val internal: InternalOrder = transform(external).to[InternalOrder]
-
-    assertEquals(
-      InternalOrder(
-        id = "order-42",
-        lines = List(InternalLine("BOOK-1", 2), InternalLine("PEN-9", 5)),
-        metadata = Map("channel" -> "web", "priority" -> "standard")
-      ),
-      internal
-    )
+    val jsonValue: ujson.Value = transform(customer).to[ujson.Value]
+    assertEquals("Dorothy Vaughan", jsonValue(objKey("customer_name")).str)
+    assertEquals(customer, transform(jsonValue).to[Customer])
   }
 
   @Test

--- a/tests/src/com.lihaoyi/upickle_3/4.4.0/user-code-filter.json
+++ b/tests/src/com.lihaoyi/upickle_3/4.4.0/user-code-filter.json
@@ -1,0 +1,13 @@
+{
+  "rules" : [
+    {
+      "excludeClasses" : "**"
+    },
+    {
+      "includeClasses" : "upickle.**"
+    },
+    {
+      "includeClasses" : "com_lihaoyi.upickle_3.**"
+    }
+  ]
+}

